p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,688 @@
+		return nil, fmt.Errorf("server must authenticate peer ID, but no peer ID provided")
+	}
+
+	httpAddrs := make([]ma.Multiaddr, 0, 1) // The common case of a single http address
+	nonHttpAddrs := make([]ma.Multiaddr, 0, len(server.Addrs))
+
+	firstAddrIsHTTP := false
+
+	for i, addr := range server.Addrs {
+		addr, isHttp := normalizeHTTPMultiaddr(addr)
+		if isHttp {
+			if i == 0 {
+				firstAddrIsHTTP = true
+			}
+			httpAddrs = append(httpAddrs, addr)
+		} else {
+			nonHttpAddrs = append(nonHttpAddrs, addr)
+		}
+	}
+
+	// Do we have an existing connection to this peer?
+	existingStreamConn := false
+	if server.ID != "" && h.streamHost != nil {
+		existingStreamConn = len(h.streamHost.Network().ConnsToPeer(server.ID)) > 0
+	}
+
+	// Currently the HTTP transport can not authenticate peer IDs.
+	if !options.ServerMustAuthenticatePeerID && len(httpAddrs) > 0 && (options.preferHTTPTransport || (firstAddrIsHTTP && !existingStreamConn)) {
+		parsed := parseMultiaddr(httpAddrs[0])
+		scheme := "http"
+		if parsed.useHTTPS {
+			scheme = "https"
+		}
+
+		rt := h.httpRoundTripper
+		ownRoundtripper := false
+		if parsed.sni != parsed.host {
+			// We have a different host and SNI (e.g. using an IP address but specifying a SNI)
+			// We need to make our own transport to support this.
+			rt = rt.Clone()
+			rt.TLSClientConfig = h.httpRoundTripper.TLSClientConfig.Clone()
+			rt.TLSClientConfig.ServerName = parsed.sni
+			ownRoundtripper = true
+		}
+
+		return &roundTripperForSpecificServer{
+			RoundTripper:     rt,
+			ownRoundtripper:  ownRoundtripper,
+			httpHost:         h,
+			server:           server.ID,
+			targetServerAddr: parsed.host + ":" + parsed.port,
+			sni:              parsed.sni,
+			scheme:           scheme,
+		}, nil
+	}
+
+	// Otherwise use a stream based transport
+	if h.streamHost == nil {
+		return nil, fmt.Errorf("can not use the HTTP transport (either no address or PeerID auth is required), and no stream host provided")
+	}
+	if !existingStreamConn {
+		if server.ID == "" {
+			return nil, fmt.Errorf("can not use the HTTP transport, and no server peer ID provided")
+		}
+		err := h.streamHost.Connect(context.Background(), peer.AddrInfo{ID: server.ID, Addrs: nonHttpAddrs})
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to peer: %w", err)
+		}
+	}
+
+	return NewStreamRoundTripper(h.streamHost, server.ID), nil
+}
+
+type httpMultiaddr struct {
+	useHTTPS bool
+	host     string
+	port     string
+	sni      string
+}
+
+func parseMultiaddr(addr ma.Multiaddr) httpMultiaddr {
+	out := httpMultiaddr{}
+	ma.ForEach(addr, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case ma.P_IP4, ma.P_IP6, ma.P_DNS, ma.P_DNS4, ma.P_DNS6:
+			out.host = c.Value()
+		case ma.P_TCP, ma.P_UDP:
+			out.port = c.Value()
+		case ma.P_TLS:
+			out.useHTTPS = true
+		case ma.P_SNI:
+			out.sni = c.Value()
+
+		}
+		return out.host == "" || out.port == "" || !out.useHTTPS || out.sni == ""
+	})
```

Comment:
Please don't reimplement multiaddr parsing. `manet.ToNetAddr` should provide half of what you need here.